checker/src/main/java/org/checkerframework/checker/collectionownership/CollectionOwnershipTransfer.java

```diff
@@ -0,0 +1,361 @@
+import org.checkerframework.dataflow.expression.JavaExpression;
+import org.checkerframework.dataflow.util.NodeUtils;
+import org.checkerframework.framework.flow.CFAbstractTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.javacutil.TreeUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+/**
+ * Transfer function for the collection ownership type system. Its primary purpose is to create
+ * temporary variables for expressions (which allow those expressions to have refined information in
+ * the store, which the consistency checker can use).
+ */
+public class CollectionOwnershipTransfer
+    extends CFAbstractTransfer<CFValue, CollectionOwnershipStore, CollectionOwnershipTransfer> {
+
+  /** The type factory. */
+  private final CollectionOwnershipAnnotatedTypeFactory atypeFactory;
+
+  /** The checker. */
+  private final CollectionOwnershipChecker checker;
+
+  /** The MustCall type factory to manage temporary variables. */
+  private final MustCallAnnotatedTypeFactory mcAtf;
+
+  /**
+   * Create a CollectionOwnershipTransfer.
+   *
+   * @param analysis the analysis
+   * @param checker the checker
+   */
+  public CollectionOwnershipTransfer(
+      CollectionOwnershipAnalysis analysis, CollectionOwnershipChecker checker) {
+    super(analysis);
+    this.atypeFactory = (CollectionOwnershipAnnotatedTypeFactory) analysis.getTypeFactory();
+    this.mcAtf = ResourceLeakUtils.getMustCallAnnotatedTypeFactory(checker);
+    this.checker = checker;
+  }
+
+  @Override
+  public TransferResult<CFValue, CollectionOwnershipStore> visitAssignment(
+      AssignmentNode node, TransferInput<CFValue, CollectionOwnershipStore> in) {
+    TransferResult<CFValue, CollectionOwnershipStore> res = super.visitAssignment(node, in);
+
+    Node lhs = getNodeOrTempVar(node.getTarget());
+    JavaExpression lhsJE = JavaExpression.fromNode(lhs);
+
+    Node rhs = getNodeOrTempVar(node.getExpression());
+    CollectionOwnershipType rhsType =
+        atypeFactory.getCoType(rhs, atypeFactory.getStoreBefore(node));
+
+    // Ownership transfer from rhs into lhs usually.
+    // Special case: desugared assignments of a temporary array variable
+    // and rhs being owning resource collection field.
+    if (rhsType != null) {
+      switch (rhsType) {
+        case OwningCollection:
+        case OwningCollectionWithoutObligation:
+          JavaExpression rhsJE = JavaExpression.fromNode(rhs);
+          if (node.isDesugaredFromEnhancedArrayForLoop()
+              || atypeFactory.isOwningCollectionField(
+                  TreeUtils.elementFromTree(node.getExpression().getTree()))) {
+            replaceInStores(res, lhsJE, atypeFactory.NOTOWNINGCOLLECTION);
+          } else {
+            replaceInStores(res, rhsJE, atypeFactory.NOTOWNINGCOLLECTION);
+          }
+          break;
+        default:
+      }
+    }
+    // boolean assignmentOfOwningCollectionArrayElement =
```

Comment:
Why is this code commented out? I assume if its not actually running here then it's not needed; in that case, why not delete it?